### PR TITLE
Add nginx support

### PR DESCRIPTION
**Pull request by yagonobre**

### CODE REVIEW

File: pkg/server/k8s/converters.go [hunk 1]
```diff
@@ -84,10 +84,13 @@ func podSpecVolumesToK8sVolumes(vols []*spec.Volume) []k8sv1.Volume {
 		vol := k8sv1.Volume{Name: v.Name}
 		if v.EmptyDir {
 			vol.EmptyDir = &k8sv1.EmptyDirVolumeSource{}
-		} else {
+		} else if v.SecretName != "" {
 			vol.Secret = &k8sv1.SecretVolumeSource{
 				SecretName: v.SecretName,
 			}
+		} else {
```
Comment (aguerra) [inline pkg/server/k8s/converters.go]: maybe to be more specific and  check for ConfigMapName

File: pkg/server/spec/container.go [hunk 2]
```diff
@@ -5,6 +5,33 @@ import (
...
+
+type VolumeMounts struct {
+	Name      string
+	MountPath string
+	ReadOnly  bool
+}
+
+type Port struct {
+	Name          string
+	ContainerPort int32
```
Comment (aguerra) [inline pkg/server/spec/container.go]: Any specific reason for using named ports or it's just a choice?

Comment (yagonobre) [reply pkg/server/spec/container.go]: I guess that if we have more than 1 one port, this need to be named.

File: pkg/server/k8s/converters.go (hunk 1 above)
Comment (yagonobre) [reply pkg/server/k8s/converters.go]: Make sense

File: pkg/server/spec/pod.go [hunk 3]
```diff
@@ -59,7 +59,7 @@ func NewPod(name, nginxImage, image string, a *app.App, envVars map[string]strin
 	ps := &Pod{
 		Name:       name,
 		Namespace:  a.Name,
-		Containers: newPodContainers(image, nginxImage, envVars, hasNginx),
+		Containers: newPodContainers(name, image, nginxImage, envVars, hasNginx),
```
[the remark this answers is not in the document]
Comment (aguerra) [reply pkg/server/spec/pod.go]: Ok, I got it, you are using `req.Name` as the container  in `LogOptions`.

File: pkg/server/spec/pod.go [hunk 4]
```diff
@@ -25,26 +25,44 @@ type Pod struct {
...
+	if hasNginx {
+		volumes = append(volumes, &Volume{
+			Name:          "nginx-conf",
+			ConfigMapName: appName,
+		})
+	}
+	return volumes
+}
+
+func newPodContainers(appImage, nginxImage string, envVars map[string]string, hasNginx bool) []*Container {
```
[the remark this answers is not in the document]
Comment (yagonobre) [reply pkg/server/spec/pod.go]: Yes, you're right, but I can't think in another solution too.

File: pkg/server/spec/pod.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (yagonobre) [reply pkg/server/spec/pod.go]: yes :D

File: pkg/server/spec/container.go [hunk 5]
```diff
@@ -36,3 +73,51 @@ func newInitContainers(slugURL, image string, a *app.App, fs storage.Storage) []
...
+	return &Container{
+		Name:  name,
+		Image: image,
+		Env:   envVars,
+		Ports: []Port{{
+			Name:          "app",
+			ContainerPort: int32(port),
+		}},
+	}
+}
```
Comment (drgarcia1986) [inline pkg/server/spec/container.go]: What about a new file for nginx specs? `spec/nginx.go1

File: pkg/server/spec/pod.go [hunk 6]
```diff
@@ -10,70 +10,72 @@ const (
...
+	c := []*Container{
+		newAppContainer(name, appImage, envVars, hasNginx),
+	}
+	if hasNginx {
+		c = append(c, newNginxContainer(nginxImage))
+	}
+	return c
+}
+
+func NewPod(name, nginxImage, image string, a *app.App, envVars map[string]string, fs storage.Storage, hasNginx bool) *Pod {
```
Comment (drgarcia1986) [inline pkg/server/spec/pod.go]: Maybe you can change args to get struct `Images` struct and check the flag `hasNginx` with `imgs.NginxImage` field.

File: pkg/server/spec/container.go (hunk 5 above)
Comment (yagonobre) [reply pkg/server/spec/container.go]: LGTM, but the nginx is just a new container spec

File: pkg/server/spec/pod.go (hunk 6 above)
Comment (yagonobre) [reply pkg/server/spec/pod.go]: Make senses

**Comment by yagonobre** [comment]
@aguerra @drgarcia1986 updated

File: pkg/server/deploy/config.go [hunk 7]
```diff
@@ -60,7 +70,7 @@ func getDeployConfigFilesFromTarBall(tarBall io.ReadSeeker, processType string)
 			return nil, err
 		}
 
-		if hdr.Name != tYamlFilename && hdr.Name != tYamlProcessTypeFileName && hdr.Name != ProcfileFileName {
+		if hdr.Name != tYamlFilename && hdr.Name != tYamlProcessTypeFileName && hdr.Name != ProcfileFileName && hdr.Name != nginxConfFileName {
```
Comment (drgarcia1986) [inline pkg/server/deploy/config.go]: What about? :
```go
if !isConfigFile(hdr.Name, tYamlFilename, tYamlProcessTypeFileName, ProcfileFileName, nginxConfFileName) {
    continue
}

// ....
func isConfigFile(name string, confFiles ...string) bool {
    for _, cf := range confFiles {
        if name == cf {
            return true
        }
        return false
}

File: pkg/server/spec/container.go (hunk 5 above)
Comment (drgarcia1986) [reply pkg/server/spec/container.go]: I know... follow your heart ❤️ :shipit:

File: pkg/server/spec/pod.go [hunk 8]
```diff
@@ -10,70 +10,82 @@ const (
...
+		c = append(c, newNginxContainer(nginxImage))
+	}
+	return c
+}
+
+func NewPod(name, nginxImage, image string, a *app.App, envVars map[string]string, fs storage.Storage) *Pod {
+	hasNginx := false
+	if nginxImage != "" {
+		hasNginx = true
+	}
```
Comment (drgarcia1986) [inline pkg/server/spec/pod.go]:
```diff
- if nginxImage != "" {
-     hasNginx = true
- }
+ hasNginx = nginxImage != ""
```

File: pkg/server/spec/pod.go [hunk 9]
```diff
@@ -10,70 +10,82 @@ const (
...
+	c := []*Container{
+		newAppContainer(name, appImage, envVars, appPort),
+	}
+	if nginxImage != "" {
+		c = append(c, newNginxContainer(nginxImage))
+	}
+	return c
+}
+
+func NewPod(name, nginxImage, image string, a *app.App, envVars map[string]string, fs storage.Storage) *Pod {
```
Comment (drgarcia1986) [inline pkg/server/spec/pod.go]: What about change `nginxImage` + `image` to `Images` struct ?

**Comment by yagonobre** [comment]
@aguerra @drgarcia1986 updated and rebased

**Pull request merged**